### PR TITLE
Migrations: Preserve entity type display name in snapshot

### LINKED ISSUE #3545: Primary Keys mangled on generic DbSets models after second migration
The migration file (i.e. 20151024123838_Init.cs) gets mangled primary keys on second migration. 

I am using AspNet Identity in my project and after the initial migration the 2nd (and following) migrations will mangle the primary keys, basically making it impossible to migrate further and having to wipe of the database after each change in the structure. 

For example: 

    public class AuthUser : IdentityUser<Guid>
    {
        public AuthUser() { }
        public AuthUser(string userName) : base(userName) { }

        public string FirstName { get; set; }
        public string LastName { get; set; }
    }
    public class AuthRole : IdentityRole<Guid>
    {
        public AuthRole() { }
        public AuthRole(string roleName) : base(roleName) { }
    }

    public class AuthContext : IdentityDbContext<AuthUser, AuthRole, Guid>
    {
    }

After setting up the initial project, run a migration 

dnx ef migrations add Init

The files genreated are generated correctly. Without further change, add a second migration

dnx ef migrations add Init2

The Init2 file (i.e.20151024123846_Init2.cs) has mangled primary keys: 

First imgration (20151024123838_Init.cs):

    public partial class Init : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "AspNetRoles",
                columns: table => new
                {
                    Id = table.Column<Guid>(nullable: false),
                    ConcurrencyStamp = table.Column<string>(nullable: true),
                    Name = table.Column<string>(nullable: true),
                    NormalizedName = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_AuthRole", x => x.Id);
                });
            migrationBuilder.CreateTable(
                name: "AspNetUsers",
                columns: table => new
                {
                    Id = table.Column<Guid>(nullable: false),
                    AccessFailedCount = table.Column<int>(nullable: false),
                    ConcurrencyStamp = table.Column<string>(nullable: true),
                    DisplayName = table.Column<string>(nullable: true),
                    Domain = table.Column<string>(nullable: true),
                    DomainUsername = table.Column<string>(nullable: true),
                    Email = table.Column<string>(nullable: true),
                    EmailConfirmed = table.Column<bool>(nullable: false),
                    FirstName = table.Column<string>(nullable: true),
                    IsDomainUser = table.Column<bool>(nullable: false),
                    LastName = table.Column<string>(nullable: true),
                    LockoutEnabled = table.Column<bool>(nullable: false),
                    LockoutEnd = table.Column<DateTimeOffset>(nullable: true),
                    NormalizedEmail = table.Column<string>(nullable: true),
                    NormalizedUserName = table.Column<string>(nullable: true),
                    PasswordHash = table.Column<string>(nullable: true),
                    PhoneNumber = table.Column<string>(nullable: true),
                    PhoneNumberConfirmed = table.Column<bool>(nullable: false),
                    SecurityStamp = table.Column<string>(nullable: true),
                    TwoFactorEnabled = table.Column<bool>(nullable: false),
                    UserName = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_AuthUser", x => x.Id);
                });
            migrationBuilder.CreateTable(
                name: "Application",
                columns: table => new
                {
                    ApplicationID = table.Column<string>(nullable: false),
                    DisplayName = table.Column<string>(nullable: true),
                    LogoutRedirectUri = table.Column<string>(nullable: true),
                    RedirectUri = table.Column<string>(nullable: true),
                    Secret = table.Column<string>(nullable: true),
                    Type = table.Column<int>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Application", x => x.ApplicationID);
                });
            migrationBuilder.CreateTable(
                name: "AspNetRoleClaims",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
                    ClaimType = table.Column<string>(nullable: true),
                    ClaimValue = table.Column<string>(nullable: true),
                    RoleId = table.Column<Guid>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_IdentityRoleClaim<Guid>", x => x.Id);
                    table.ForeignKey(
                        name: "FK_IdentityRoleClaim<Guid>_AuthRole_RoleId",
                        column: x => x.RoleId,
                        principalTable: "AspNetRoles",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                });
            migrationBuilder.CreateTable(
                name: "AspNetUserClaims",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
                    ClaimType = table.Column<string>(nullable: true),
                    ClaimValue = table.Column<string>(nullable: true),
                    UserId = table.Column<Guid>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_IdentityUserClaim<Guid>", x => x.Id);
                    table.ForeignKey(
                        name: "FK_IdentityUserClaim<Guid>_AuthUser_UserId",
                        column: x => x.UserId,
                        principalTable: "AspNetUsers",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                });
            migrationBuilder.CreateTable(
                name: "AspNetUserLogins",
                columns: table => new
                {
                    LoginProvider = table.Column<string>(nullable: false),
                    ProviderKey = table.Column<string>(nullable: false),
                    ProviderDisplayName = table.Column<string>(nullable: true),
                    UserId = table.Column<Guid>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_IdentityUserLogin<Guid>", x => new { x.LoginProvider, x.ProviderKey });
                    table.ForeignKey(
                        name: "FK_IdentityUserLogin<Guid>_AuthUser_UserId",
                        column: x => x.UserId,
                        principalTable: "AspNetUsers",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                });
            migrationBuilder.CreateTable(
                name: "AspNetUserRoles",
                columns: table => new
                {
                    UserId = table.Column<Guid>(nullable: false),
                    RoleId = table.Column<Guid>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_IdentityUserRole<Guid>", x => new { x.UserId, x.RoleId });
                    table.ForeignKey(
                        name: "FK_IdentityUserRole<Guid>_AuthRole_RoleId",
                        column: x => x.RoleId,
                        principalTable: "AspNetRoles",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                    table.ForeignKey(
                        name: "FK_IdentityUserRole<Guid>_AuthUser_UserId",
                        column: x => x.UserId,
                        principalTable: "AspNetUsers",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                });
            migrationBuilder.CreateIndex(
                name: "RoleNameIndex",
                table: "AspNetRoles",
                column: "NormalizedName");
            migrationBuilder.CreateIndex(
                name: "EmailIndex",
                table: "AspNetUsers",
                column: "NormalizedEmail");
            migrationBuilder.CreateIndex(
                name: "UserNameIndex",
                table: "AspNetUsers",
                column: "NormalizedUserName");
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropTable("AspNetRoleClaims");
            migrationBuilder.DropTable("AspNetUserClaims");
            migrationBuilder.DropTable("AspNetUserLogins");
            migrationBuilder.DropTable("AspNetUserRoles");
            migrationBuilder.DropTable("Application");
            migrationBuilder.DropTable("AspNetRoles");
            migrationBuilder.DropTable("AspNetUsers");
        }
    }

Second, mangeld migration (20151024123846_Init2.cs):

    public partial class Init2 : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropForeignKey(name: "FK_Guid>_AuthRole_RoleId", table: "AspNetRoleClaims");
            migrationBuilder.DropForeignKey(name: "FK_Guid>_AuthUser_UserId", table: "AspNetUserClaims");
            migrationBuilder.DropForeignKey(name: "FK_Guid>_AuthUser_UserId", table: "AspNetUserLogins");
            migrationBuilder.DropForeignKey(name: "FK_Guid>_AuthRole_RoleId", table: "AspNetUserRoles");
            migrationBuilder.DropForeignKey(name: "FK_Guid>_AuthUser_UserId", table: "AspNetUserRoles");
            migrationBuilder.DropPrimaryKey(name: "PK_Guid>", table: "AspNetUserRoles");
            migrationBuilder.DropPrimaryKey(name: "PK_Guid>", table: "AspNetUserLogins");
            migrationBuilder.DropPrimaryKey(name: "PK_Guid>", table: "AspNetUserClaims");
            migrationBuilder.DropPrimaryKey(name: "PK_Guid>", table: "AspNetRoleClaims");
            migrationBuilder.AddPrimaryKey(
                name: "PK_IdentityUserRole<Guid>",
                table: "AspNetUserRoles",
                columns: new[] { "UserId", "RoleId" });
            migrationBuilder.AddPrimaryKey(
                name: "PK_IdentityUserLogin<Guid>",
                table: "AspNetUserLogins",
                columns: new[] { "LoginProvider", "ProviderKey" });
            migrationBuilder.AddPrimaryKey(
                name: "PK_IdentityUserClaim<Guid>",
                table: "AspNetUserClaims",
                column: "Id");
            migrationBuilder.AddPrimaryKey(
                name: "PK_IdentityRoleClaim<Guid>",
                table: "AspNetRoleClaims",
                column: "Id");
            migrationBuilder.AddForeignKey(
                name: "FK_IdentityRoleClaim<Guid>_AuthRole_RoleId",
                table: "AspNetRoleClaims",
                column: "RoleId",
                principalTable: "AspNetRoles",
                principalColumn: "Id",
                onDelete: ReferentialAction.Cascade);
            migrationBuilder.AddForeignKey(
                name: "FK_IdentityUserClaim<Guid>_AuthUser_UserId",
                table: "AspNetUserClaims",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Cascade);
            migrationBuilder.AddForeignKey(
                name: "FK_IdentityUserLogin<Guid>_AuthUser_UserId",
                table: "AspNetUserLogins",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Cascade);
            migrationBuilder.AddForeignKey(
                name: "FK_IdentityUserRole<Guid>_AuthRole_RoleId",
                table: "AspNetUserRoles",
                column: "RoleId",
                principalTable: "AspNetRoles",
                principalColumn: "Id",
                onDelete: ReferentialAction.Cascade);
            migrationBuilder.AddForeignKey(
                name: "FK_IdentityUserRole<Guid>_AuthUser_UserId",
                table: "AspNetUserRoles",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Cascade);
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropForeignKey(name: "FK_IdentityRoleClaim<Guid>_AuthRole_RoleId", table: "AspNetRoleClaims");
            migrationBuilder.DropForeignKey(name: "FK_IdentityUserClaim<Guid>_AuthUser_UserId", table: "AspNetUserClaims");
            migrationBuilder.DropForeignKey(name: "FK_IdentityUserLogin<Guid>_AuthUser_UserId", table: "AspNetUserLogins");
            migrationBuilder.DropForeignKey(name: "FK_IdentityUserRole<Guid>_AuthRole_RoleId", table: "AspNetUserRoles");
            migrationBuilder.DropForeignKey(name: "FK_IdentityUserRole<Guid>_AuthUser_UserId", table: "AspNetUserRoles");
            migrationBuilder.DropPrimaryKey(name: "PK_IdentityUserRole<Guid>", table: "AspNetUserRoles");
            migrationBuilder.DropPrimaryKey(name: "PK_IdentityUserLogin<Guid>", table: "AspNetUserLogins");
            migrationBuilder.DropPrimaryKey(name: "PK_IdentityUserClaim<Guid>", table: "AspNetUserClaims");
            migrationBuilder.DropPrimaryKey(name: "PK_IdentityRoleClaim<Guid>", table: "AspNetRoleClaims");
            migrationBuilder.AddPrimaryKey(
                name: "PK_Guid>",
                table: "AspNetUserRoles",
                columns: new[] { "UserId", "RoleId" });
            migrationBuilder.AddPrimaryKey(
                name: "PK_Guid>",
                table: "AspNetUserLogins",
                columns: new[] { "LoginProvider", "ProviderKey" });
            migrationBuilder.AddPrimaryKey(
                name: "PK_Guid>",
                table: "AspNetUserClaims",
                column: "Id");
            migrationBuilder.AddPrimaryKey(
                name: "PK_Guid>",
                table: "AspNetRoleClaims",
                column: "Id");
            migrationBuilder.AddForeignKey(
                name: "FK_Guid>_AuthRole_RoleId",
                table: "AspNetRoleClaims",
                column: "RoleId",
                principalTable: "AspNetRoles",
                principalColumn: "Id",
                onDelete: ReferentialAction.Restrict);
            migrationBuilder.AddForeignKey(
                name: "FK_Guid>_AuthUser_UserId",
                table: "AspNetUserClaims",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Restrict);
            migrationBuilder.AddForeignKey(
                name: "FK_Guid>_AuthUser_UserId",
                table: "AspNetUserLogins",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Restrict);
            migrationBuilder.AddForeignKey(
                name: "FK_Guid>_AuthRole_RoleId",
                table: "AspNetUserRoles",
                column: "RoleId",
                principalTable: "AspNetRoles",
                principalColumn: "Id",
                onDelete: ReferentialAction.Restrict);
            migrationBuilder.AddForeignKey(
                name: "FK_Guid>_AuthUser_UserId",
                table: "AspNetUserRoles",
                column: "UserId",
                principalTable: "AspNetUsers",
                principalColumn: "Id",
                onDelete: ReferentialAction.Restrict);
        }
    }

The `dnx ef database update` will now of course fail.

As you can see "FK_IdentityRoleClaim<Guid>_AuthRole_RoleId" got mangled into "FK_Guid>_AuthRole_RoleId" and the "IdentityRoleClaim<" was removed. 

It's a big game breaker, as migrations stop working when generic models are used.

### PR DESCRIPTION
**Pull request by bricelam**
This also removes any semantics about a shadow entity names.

Fixes #3545

@smitpatel

### CODE REVIEW

File: src/EntityFramework.Commands/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 1]
```diff
@@ -335,7 +335,14 @@ protected virtual void GenerateEntityTypeAnnotations([NotNull] IEntityType entit
             Check.NotNull(entityType, nameof(entityType));
             Check.NotNull(stringBuilder, nameof(stringBuilder));
 
-            var annotations = entityType.GetAnnotations().ToArray();
+            var annotations = entityType.GetAnnotations().ToList();
+
+            var displayName = entityType.DisplayName();
```
Comment (smitpatel) [inline src/EntityFramework.Commands/Migrations/Design/CSharpSnapshotGenerator.cs]: worth putting a comment for this?

**Comment by smitpatel** [comment]
:shipit:

**Comment by bricelam** [comment]
Hmm, I hate that this adds an annotation to virtually *every* entity type in the snapshot...

Any better ideas @ajcvickers @AndriySvyryd @anpete @divega @rowanmiller?

The root of the problem is that the default table name (and thus the key, index, and foreign key names) are based off `EntityType.DisplayName()` which in turn is based off `ClrType` (when available) or `Name`. For "unbound" types, this means that we need to convert Names like `My.Ugly+Type<System.Guid, My.Ugly+Struct<int>>` into `Type<Guid, Struct<int>>` which requires string parsing and adds unnecessary semantics to `EntityType.Name`.

This PR takes away the semantics and captures `DisplayName` in the snapshot for "unbound" entity types. This removes the need to parse, but adds extra stuff to the snapshot.

File: src/EntityFramework.Core/Metadata/Internal/EntityType.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ public EntityType([NotNull] string name, [NotNull] Model model, ConfigurationSou
 
             _properties = new SortedDictionary<string, Property>(new PropertyComparer(this));
 #if DEBUG
-            DebugName = DisplayName();
+            DebugName = EntityTypeExtensions.DisplayName(this);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/EntityType.cs]: this.DisplayName() ?

**Comment by anpete** [comment]
I'm not sure I fully understand this, can we discuss in person?

File: src/EntityFramework.Core/Metadata/Internal/EntityType.cs (hunk 2 above)
Comment (bricelam) [inline src/EntityFramework.Core/Metadata/Internal/EntityType.cs]: Which is the lesser of two evils..?

**Comment by ajcvickers** [comment]
@bricelam This is a consequence of having shadow state entity types in the snapshot, right? In other words, the model used to create the snapshot has real entity types, but the snapshot model loads these as shadow types, which then causes a difference in behavior with regards to names.

**Comment by bricelam** [comment]
@ajcvickers Exactly. By converting CLR-backed entity types to shadow entities, we loose some information used by `DisplayName`.

**Comment by anpete** [comment]
Could we write the table name annotation instead?

**Comment by bricelam** [comment]
@anpete Yes, but we'd also have to update constraint names to be based on that (instead of entity type name) which means when you rename an entity type, all the constraints touching it would also be renamed resulting in a larger diff than expected.

**Comment by ajcvickers** [comment]
@bricelam Could we annotate with the display name only for generic types?

File: src/EntityFramework.Core/Metadata/Internal/EntityType.cs (hunk 2 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/EntityType.cs]: +1 for this

**Comment by anpete** [comment]
And possibly we should call the annotation something else

**Comment by bricelam** [comment]
@ajcvickers The existing code has a few other quarks that I don't like. For example, if I name a shadow entity `Post+Contributor`, the default table name would be `Contributor` because we apply CLR type semantics to the name.

**Comment by anpete** [comment]
@bricelam Could we have a simpler way of producing a table name from a type name string? I.e. not use DisplayName?

**Comment by bricelam** [comment]
This PR simplifies it to...

1. Use `ClrType.DisplayName()` (a pretty version of `Type.Name`)
2. If no `ClrType` use `DisplayName` annotation
3. If no annotation, use `Name` verbatim.

**Comment by bricelam** [comment]
We could change `EntityType.Name`... E.g. To trip out generics, but this would prevent you from adding the entity types `Thing<int>` and `Thing<string>` to the same model.

**Comment by bricelam** [comment]
We could also strip out the namespaces of the generic type arguments which would simplify the conversion:
Name: `My.Thing<Guid>` (instead of `My.Thing<System.Guid>` like it currently is)
TableName: `Thing<Guid>` (same as it currently is)

**Comment by bricelam** [comment]
Recalling. We decided to serialize `TableName` instead of an annotation and change the constraint names to be based on table name instead of entity type name.

**Pull request closed without merging**